# **Naive Bayes**

## **Apa itu Naive Bayes?**
Naïve bayes atau dikenal juga dengan naïve bayes classifier merupakan salah satu algoritme machine learning yang diawasi (supervised learning) yang digunakan untuk menangani masalah klasifikasi berdarkan pada probabilitas atau kemungkinan sesuai dengan Teorema Bayes.

Teorema Bayes atau hukum bayes menjelaskan probabilitas suatu kejadian di masa depan berdasarkan pengalaman sebelumnya tentang kondisi yang mungkin terkait dengan kejadian tersebut, sehingga dapat digunakan untuk pengambilan keputusan. Contoh, terdapat 10 kartu bernomor 1 sampai 10. Jika satu kartu diambil secara acak, maka peluang mendapatkan kartu bernomor 6 adalah 1/10. Angka 1 didapatkan dari kemungkinan kartu bernomor 6 diambil, sedangkan angka 10 merupakan total keseluruhan dari kartu yang ada.

Algoritme Naïve bayes disebut “naif” karena membuat asumsi yang sangat kuat (naif) bahwa kemunculan fitur tertentu tidak tergantung pada kemunculan fitur lainnya (kelas). Misal, pemohon pinjaman (kredit) disetujui atau tidak, tergantung pada pendapatan per bulan, riwayat pinjaman dan transaksi sebelumnya, usia serta lokasi. Meski fitur-fitur tersebut saling bergantung satu sama lain, namun semua fitur-fitur tersebut secara individual memiliki kontribusi pada kemungkinan apakah permohonan kredit seseorang disetujui atau tidak.

### **Contoh implementasi algoritme naïve bayes**
- klasifikasi dokumen
- perkiraan cuaca
- mendeteksi atau menyaring spam
- system rekomendasi
- analisis sentiment

---
## **Metode Naive bayes**

1. Multinomial naïve bayes

    metode naïve bayes yang digunakan untuk mengklasifikasikan kategori dokumen. Misalnya, sebuah dokumen diklasifikasikan bertema teknologi, olahraga, politik dan lain sebagainya berdasarkan frekuensi kata-kata yang muncul di dalam dokumen tersebut.

2. Gaussian naïve bayes

    tipe naïve bayes yang mendukung data kontinu (tidak terbatas) berdasarkan asumsi ditribusi gaussian.

3. Bernoulli naïve bayes

    Metode ini hampir sama seperti tipe multinomial, bedanya, tipe Bernoulli lebih berfokus pada hasil yang bernilai Boolean yaitu benar atau salah. Contoh, prediksi apakah sebuah kata tertentu muncul dalam dokumen teks atau tidak.

---

## **Prinsip dasar Teorama Bayes**

**Teorema Bayes**:


$
P(Y|X) = \frac{P(X|Y) \cdot P(Y)}{P(X)}
$

**Penjelasan**:

- \( P(Y|X) \): Probabilitas posterior (kelas \( Y \) diberikan fitur \( X \)).
- \( P(X|Y) \): Likelihood (probabilitas fitur \( X \) diberikan kelas \( Y \)).
- \( P(Y) \): Prior probability (probabilitas awal kelas \( Y \)).
- \( P(X) \): Evidence (probabilitas fitur \( X \) secara keseluruhan).

**Tujuan**:

Mencari kelas \( Y \) dengan nilai \( P(Y|X) \) tertinggi.

---

## **Langkah-langkah Klasifikasi**
1. Persiapan Data:

    Pisahkan data menjadi fitur (𝑋) dan label (𝑌).
2. Hitung Probabilitas Awal (𝑃(𝑌)):

    Hitung frekuensi setiap kelas dalam dataset.
3. Hitung Likelihood (𝑃(𝑋|𝑌)):

    Untuk setiap fitur, hitung probabilitas kemunculannya dalam setiap kelas.
4. Gabungkan dengan Teorema Bayes:
    
    Gunakan rumus

    $
    P(Y|X) = \frac{P(X|Y) \cdot P(Y)}{P(X)}
    $
5. Prediksi Kelas:
    
    Pilih kelas dengan nilai 𝑃(𝑌|𝑋) tertinggi.


---

## **Contoh Kasus Naïve Bayes untuk Data Numerik**

### **Dataset Awal**

Misalkan kita memiliki dataset tentang deteksi kecurangan transaksi berdasarkan dua fitur numerik:

| Fitur 1 (Jumlah Transaksi dalam Sehari) | Fitur 2 (Nominal Transaksi Rata-rata) | Label (Kelas) |
|----------------------------------------|--------------------------------|-------------|
| 15                                     | 500.000                        | Fraud      |
| 3                                      | 150.000                        | Legit      |
| 20                                     | 700.000                        | Fraud      |
| 2                                      | 100.000                        | Legit      |
| 12                                     | 450.000                        | Fraud      |



### **Langkah-Langkah Perhitungan**

#### **Langkah 1: Hitung Probabilitas Prior**
- Jumlah total data: $ N = 5 $
- Jumlah transaksi fraud: $ N_{fraud} = 3 $
- Jumlah transaksi legit: $ N_{legit} = 2 $

Probabilitas prior:

$
P(Fraud) = \frac{3}{5} = 0.6
$

$
P(Legit) = \frac{2}{5} = 0.4
$



#### **Langkah 2: Hitung Mean dan Varians untuk Setiap Fitur**
Untuk setiap kelas (Fraud dan Legit), kita hitung **mean ($\mu$)** dan **varians ($\sigma^2$)** dari setiap fitur.

##### **Fitur 1: Jumlah Transaksi dalam Sehari**
**Kelas Fraud**: Data = {15, 20, 12}  

$
\mu_{fraud} = \frac{15 + 20 + 12}{3} = 15.67
$
$
\sigma^2_{fraud} = \frac{(15-15.67)^2 + (20-15.67)^2 + (12-15.67)^2}{3} = 16.33
$

**Kelas Legit**: Data = {3, 2}  

$
\mu_{legit} = \frac{3 + 2}{2} = 2.5
$
$
\sigma^2_{legit} = \frac{(3-2.5)^2 + (2-2.5)^2}{2} = 0.25
$



##### **Fitur 2: Nominal Transaksi Rata-rata**
**Kelas Fraud**: Data = {500.000, 700.000, 450.000}  

$
\mu_{fraud} = \frac{500000 + 700000 + 450000}{3} = 550000
$

$
\sigma^2_{fraud} = \frac{(500000 - 550000)^2 + (700000 - 550000)^2 + (450000 - 550000)^2}{3} = 16666666667
$

**Kelas Legit**: Data = {150.000, 100.000}  

$
\mu_{legit} = \frac{150000 + 100000}{2} = 125000
$
$
\sigma^2_{legit} = \frac{(150000 - 125000)^2 + (100000 - 125000)^2}{2} = 625000000
$



#### **Langkah 3: Gunakan Distribusi Gaussian**
Fungsi distribusi normal (Gaussian) untuk likelihood:

$
P(X|Y) = \frac{1}{\sqrt{2\pi\sigma^2}} \exp\left(-\frac{(X - \mu)^2}{2\sigma^2} \right)
$



#### **Langkah 4: Prediksi untuk Data Baru**
Misalkan ada transaksi baru dengan:
- **Jumlah transaksi dalam sehari** = **10**
- **Nominal transaksi rata-rata** = **400.000**

Kita hitung probabilitas untuk kelas **Fraud** dan **Legit**.

##### **Likelihood untuk Fraud**

$
P(10 | Fraud) = \frac{1}{\sqrt{2\pi(16.33)}} \exp\left(-\frac{(10 - 15.67)^2}{2(16.33)} \right)
$

$
= \frac{1}{\sqrt{2\pi(16.33)}} \exp\left(-\frac{(-5.67)^2}{32.66} \right)
$

$
= \frac{1}{\sqrt{102.53}} \exp\left(-\frac{32.15}{32.66} \right)
$

$
= \frac{1}{10.12} \exp(-0.98)
$

$
= 0.099 \times 0.375 = 0.0371
$

$
P(400000 | Fraud) = \frac{1}{\sqrt{2\pi(16666666667)}} \exp\left(-\frac{(400000 - 550000)^2}{2(16666666667)} \right)
$

$
= \frac{1}{\sqrt{1.05 \times 10^{11}}} \exp\left(-\frac{(-150000)^2}{3.33 \times 10^{10}} \right)
$

$
= \frac{1}{3.24 \times 10^5} \exp(-0.675)
$

$
= (3.08 \times 10^{-6}) \times (0.509) = 1.57 \times 10^{-6}
$



##### **Likelihood untuk Legit**

$
P(10 | Legit) = \frac{1}{\sqrt{2\pi(0.25)}} \exp\left(-\frac{(10 - 2.5)^2}{2(0.25)} \right)
$

$
= \frac{1}{\sqrt{1.57}} \exp\left(-\frac{(7.5)^2}{0.5} \right)
$

$
= \frac{1}{1.25} \exp(-56.25)
$

$
\approx 0
$

$
P(400000 | Legit) = \frac{1}{\sqrt{2\pi(625000000)}} \exp\left(-\frac{(400000 - 125000)^2}{2(625000000)} \right)
$

$
= \frac{1}{\sqrt{3.93 \times 10^9}} \exp\left(-\frac{(275000)^2}{1.25 \times 10^9} \right)
$

$
= \frac{1}{62720} \exp(-6.05)
$

$
\approx (1.59 \times 10^{-5}) \times (0.00236) = 3.76 \times 10^{-8}
$



#### **Langkah 5: Hitung Probabilitas Posterior**

$
P(Fraud | X) \propto P(10 | Fraud) \times P(400000 | Fraud) \times P(Fraud)
$

$
= 0.0371 \times 1.57 \times 10^{-6} \times 0.6
$

$
= 3.49 \times 10^{-8}
$

$
P(Legit | X) \propto P(10 | Legit) \times P(400000 | Legit) \times P(Legit)
$

$
= 0 \times 3.76 \times 10^{-8} \times 0.4 = 0
$



#### **Kesimpulan**
Karena $ P(Fraud | X) > P(Legit | X) $, transaksi baru diklasifikasikan sebagai **Fraud**.

---

## **Membandingkan akurasi pada data aris dengan outlier dan tanpa outlier menggunakan naive bayes.**

upload data iris.csv:

In [1]:
from google.colab import files

# Unggah file CSV
uploaded = files.upload()


Saving df_combined_outlier.csv to df_combined_outlier.csv


### **Import Library yagn diperlukan**

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.neighbors import LocalOutlierFactor


**tampilkan dataset**

In [4]:
# Baca dataset
df = pd.read_csv(list(uploaded.keys())[0])

# Tampilkan 5 baris pertama
df.head()

,id,Class,petal length,petal width,sepal length,sepal width
0,1,Iris-setosa,86.4,70.0,20.1,30.5
1,2,Iris-setosa,1.4,0.2,4.9,3.0
2,3,Iris-setosa,1.3,0.2,4.7,3.2
3,4,Iris-setosa,1.5,0.2,4.6,3.1
4,5,Iris-setosa,1.4,0.2,5.0,3.6


**Preprocessing Data**

In [5]:
#  Hapus kolom 'id' jika ada
if 'id' in df.columns:
    df.drop(columns=['id'], inplace=True)

#  Konversi label 'Class' ke numerik
df['Class'] = df['Class'].astype('category').cat.codes

# Pisahkan fitur dan target
X = df.drop(columns=['Class'])
y = df['Class']


**Split data**

In [7]:
# Split data 80% training, 20% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


**Latih Model Naïve Bayes dengan Outlier**

In [8]:
#  Inisialisasi model Naïve Bayes
nb = GaussianNB()

#  Latih model dengan data yang mengandung outlier
nb.fit(X_train, y_train)

#  Prediksi
y_pred = nb.predict(X_test)

#  Hitung akurasi
accuracy_with_outliers = accuracy_score(y_test, y_pred)
print(f"Akurasi dengan outlier: {accuracy_with_outliers:.4f}")


Akurasi dengan outlier: 0.9667


**Menghapus Outlier dengan LOF**

In [9]:
# Inisialisasi Local Outlier Factor (LOF)
lof = LocalOutlierFactor(n_neighbors=3)

#  Prediksi outlier (-1 = outlier, 1 = inlier)
outlier_labels = lof.fit_predict(X)

#  Hapus outlier
df_no_outlier = df[outlier_labels == 1]

#  Pisahkan kembali fitur dan target setelah menghapus outlier
X_no_outlier = df_no_outlier.drop(columns=['Class'])
y_no_outlier = df_no_outlier['Class']

#  Split data tanpa outlier
X_train_no, X_test_no, y_train_no, y_test_no = train_test_split(X_no_outlier, y_no_outlier, test_size=0.2, random_state=42)


**Latih Model Naïve Bayes Tanpa Outlier**

In [10]:
#  Latih model tanpa outlier
nb.fit(X_train_no, y_train_no)

#  Prediksi
y_pred_no = nb.predict(X_test_no)

#  Hitung akurasi tanpa outlier
accuracy_without_outliers = accuracy_score(y_test_no, y_pred_no)
print(f"Akurasi tanpa outlier: {accuracy_without_outliers:.4f}")


Akurasi tanpa outlier: 0.8929


In [ ]:
# Perbandingan hasil
print(f"Akurasi dengan outlier    : {accuracy_with_outliers:.4f}")
print(f"Akurasi tanpa outlier     : {accuracy_without_outliers:.4f}")

if accuracy_without_outliers > accuracy_with_outliers:
    print(" Menghapus outlier MENINGKATKAN akurasi.")
elif accuracy_without_outliers < accuracy_with_outliers:
    print(" Menghapus outlier MENURUNKAN akurasi.")
else:
    print(" Menghapus outlier TIDAK BERDAMPAK pada akurasi.")


Akurasi dengan outlier    : 0.9667
Akurasi tanpa outlier     : 0.8929
 Menghapus outlier MENURUNKAN akurasi.


## Kesimpulan

Setelah melakukan analisis **Naïve Bayes** pada dataset **Iris**, hasil perbandingan akurasi adalah sebagai berikut:

| **Kondisi Data**       | **Akurasi Naïve Bayes** |
|------------------------|------------------------|
| **Dengan Outlier**     |  **96.67%**           |
| **Tanpa Outlier (LOF)** |  **89.29%**           |



###  **Apa yang Terjadi?**
- Setelah **menghapus outlier**, akurasi **justru menurun** dari **96.67% menjadi 89.29%**.
- Hal ini menunjukkan bahwa **data yang dianggap outlier sebenarnya memiliki informasi penting** yang membantu model dalam melakukan klasifikasi.
- Outlier dalam dataset Iris kemungkinan **bukan anomali yang merusak**, melainkan bagian dari pola alami dalam data.


